builder/cloudstack/step_create_instance.go

```diff
@@ -65,7 +115,29 @@ func (s *stepCreateInstance) Run(state multistep.StateBag) multistep.StepAction
 	}
 
 	if config.UserData != "" {
-		ud, err := getUserData(config.UserData, config.HTTPGetOnly)
+		var httpPort uint = state.Get("http_port").(uint)
+		hostIp, err := hostIP()
+		if err != nil {
+			state.Put("error", err)
+			ui.Error(err.Error())
+			return multistep.ActionHalt
+		}
+		common.SetHTTPIP(hostIp)
+
+		s.Ctx.Data = &createInstanceTemplateData{
+			hostIp,
+			httpPort,
+		}
+
+		renderedUserData, err := interpolate.Render(config.UserData, &s.Ctx)
+		if err != nil {
+			err := fmt.Errorf("Error rendering user_data: %s", err)
+			state.Put("error", err)
+			ui.Error(err.Error())
+			return multistep.ActionHalt
+		}
+
+		ud, err := getUserData(renderedUserData, config.HTTPGetOnly)
```

Comment:
Maybe we should refactor this to `ud, err := s.generateUserData(config.UserData, config.HTTPGetOnly)` and move all the additional logic you added here to that function? It would make this function a lot cleaner and it would make the error handling a little less verbose. What do you think?